File: android/src/main/java/com/parse/flutterparse/FlutterParsePlugin.java
package com.parse.flutterparse;

import com.parse.FlutterParse;
import com.parse.FlutterParseInstallation;
import com.parse.FlutterParseObject;
import com.parse.FlutterParseQuery;
import com.parse.FlutterParseUser;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

public class FlutterParsePlugin implements MethodCallHandler {
  private final Registrar registrar;

  private FlutterParsePlugin(Registrar registrar) {
    this.registrar = registrar;
  }

  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "flutter_parse");
    channel.setMethodCallHandler(new FlutterParsePlugin(registrar));
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    switch (call.method) {
      case FlutterParse.INITIALIZE:
        FlutterParse.initialize(registrar.activeContext().getApplicationContext(), call, result);
        break;
      case FlutterParseInstallation.INSTALLATION:
        FlutterParseInstallation.init(result);
        break;
      case FlutterParseUser.GET_CURRENT_USER:
        FlutterParseUser.getCurrentUser(result);
        break;
      case FlutterParseUser.LOGIN:
        FlutterParseUser.login(call, result);
        break;
      case FlutterParseUser.REGISTER:
        FlutterParseUser.register(call, result);
        break;
      case FlutterParseUser.LOGOUT:
        FlutterParseUser.logout(result);
        break;
      case FlutterParseObject.DELETE_IN_BACKGROUND:
        FlutterParseObject.deleteAsync(call, result, false);
        break;
      case FlutterParseObject.DELETE_EVENTUALLY:
        FlutterParseObject.deleteAsync(call, result, true);
        break;
      case FlutterParseObject.FETCH_IN_BACKGROUND:
        FlutterParseObject.fetchAsync(call, result);
        break;
      case FlutterParseObject.SAVE_IN_BACKGROUND:
        FlutterParseObject.saveAsync(call, result, false);
        break;
      case FlutterParseObject.SAVE_EVENTUALLY:
        FlutterParseObject.saveAsync(call, result, true);
        break;
      case FlutterParseQuery.QUERY_IN_BACKGROUND:
        FlutterParseQuery.executeAsync(call, result);
        break;
      default:
        result.notImplemented();
        break;
    }
  }
}

File: ios/Classes/FlutterParsePlugin.m
#import "FlutterParsePlugin.h"
#import "FlutterParse.h"
#import "FlutterParseObject.h"
#import "FlutterParseUser.h"
#import "FlutterParseQuery.h"

@implementation FlutterParsePlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel methodChannelWithName:@"flutter_parse"
                                                                binaryMessenger:[registrar messenger]];
    FlutterParsePlugin* instance = [[FlutterParsePlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    // Parse
    if ([kParseInitialize isEqualToString:call.method]) {
        [FlutterParse initialize:call result:result];
        return;
    } else
    // Parse Object
    if ([kParseObjectSaveInBackground isEqualToString:call.method]) {
        [FlutterParseObject saveAsync:call result:result eventually:NO];
    } else if ([kParseObjectSaveEventually isEqualToString:call.method]) {
        [FlutterParseObject saveAsync:call result:result eventually:YES];
    } else if ([kParseObjectDeleteInBackground isEqualToString:call.method]) {
        [FlutterParseObject deleteAsync:call result:result eventually:NO];
    } else if ([kParseObjectDeleteEventually isEqualToString:call.method]) {
        [FlutterParseObject deleteAsync:call result:result eventually:YES];
    } else if ([kParseObjectFetchInBackground isEqualToString:call.method]) {
        [FlutterParseObject fetchAsync:call result:result];
    } else
    // Parse User
    if ([kParseUserLogin isEqualToString:call.method]) {
        [FlutterParseUser login:call result:result];
    } else if ([kParseUserLogout isEqualToString:call.method]) {
        [FlutterParseUser logoutWithResult:result];
    } else if ([kParseUserRegister isEqualToString:call.method]) {
        [FlutterParseUser signup:call result:result];
    } else if ([kParseUserGetCurrentUser isEqualToString:call.method]) {
        [FlutterParseUser getCurrentUserWithResult:result];
    } else
    // Parse Query
    if ([kParseQueryInBackground isEqualToString:call.method]) {
        [FlutterParseQuery executeAsync:call result:result];
    }
    
    else {
        result(FlutterMethodNotImplemented);
    }
}

@end
